### ISSUE TITLE
[QST] Would the new Arrow C Data Interface be useful for CuDF?

### ISSUE THREAD

**Issue by andygrove** [rapidsai/cudf#5097]
**What is your question?**

Apache Arrow has released a new C Data Interface [1] and it seems like it might potentially be of interest to the CuDF project, so I wanted to make sure that people are at least aware of this.

Here is an extract from the blog post that seems relevant.

> However, some projects may wish to import and export the Arrow columnar format without taking on a new library dependency, such as the Arrow C++ library. We have therefore designed an alternative which exchanges data at the C level, conforming to a simple data definition. The C Data Interface carries no dependencies except a shared C ABI between binaries which use it. C ABIs are platform-wide standards which are necessarily adhered to by all compilers which generate binaries and are extremely stable, ensuring portability of libraries and executable binaries. Two libraries that utilize the C structures defined by the C Data Interface can do zero-copy data transfers at runtime without any build-time or link-time dependency requirements.

[1] https://arrow.apache.org/blog/2020/05/03/introducing-arrow-c-data-interface/

**Comment by kkraus14** [comment]
Documentation on the interface: https://arrow.apache.org/docs/format/CDataInterface.html

**Comment by kkraus14** [comment]
Big concern from me is there's no way to differentiate between GPU and host memory via this interface currently. This means if we exposed this interface with GPU memory pointers, a downstream consumer could pick it up assuming it's host memory and blow up because of it.

While we can use built in CUDA features to check whether a given pointer is host or device memory, non-GPU libraries wouldn't perform these checks. We should work with the community to extend the interface to allow differentiating between host and device memory.

**Comment by fsaintjacques** [comment]
This is meant to pass data around between libraries/runtimes without having to link with Arrow's C++ (and copying). The first consumer in the wild of this interface is to pass data from R to Python without copying in the same process.

I would say that implementing this feature scoped to host memory only (both in and out) is already a major milestone, essentially dispatching some computation to the GPU an getting results back.

**Comment by kkraus14** [comment]
> I would say that implementing this feature scoped to host memory only (both in and out) is already a major milestone, essentially dispatching some computation to the GPU an getting results back.

So all of our data structures are ultimately backed by GPU memory where for consuming the interface it makes sense where we can copy it to the GPU and then immediately call the release callback once the copy is done.

In the other direction, we'd need to allocate host memory to copy the GPU memory into and then manage the lifetime of those memory buffers via the release callback.

**Comment by fsaintjacques** [comment]
> In the other direction, we'd need to allocate host memory to copy the GPU memory into and then manage the lifetime of those memory buffers via the release callback.

This would not require much effort since cudf already depends on libarrow, just use the C [bridge](https://github.com/apache/arrow/blob/master/cpp/src/arrow/c/bridge.h) around a Table/RecordBatch. All that is required, is to implement a cudf::table to Arrow::Table.


Also note, that Antoine recently [introduced](https://github.com/apache/arrow/commit/9f0c70c8337b1a8c75df5cc9410b54e97351bfc0) a MemoryManager/Device concept. Maybe this can help in the conversion?

**Comment by kkraus14** [comment]
cc @trxcllnt as I imagine you may have interest in implementing this 😃

**Comment by felipeblazing** [comment]
Blazing looked into this a while back. A long while back. Part of the issue then was the pain in changing some of the encodings. Strings used to be encoded very differently. Not so much anymore. The other issue was different ways in how we managed certain data types  like datetime though I think @trxcllnt might have changed some of that if I remember correctly. 

Another issue is that right now a cudf column is really just an arrow::buffer not an arrow column in the sense that they cannot be chunked in cudf the way it can in arrow. This is handled with dask in python and blazing handles chunking because our execution graph in c++ manages this and then the user just gets a dask_cudf. With this limitation it is not always true that an arrow::table can "fit" into a cudf::table because the arrow::table can contain more than 2^31 values and the cudf::table cant. (I am pretty sure about an arrow::table being able to be bigger but I am NO arrow expert.

There have been other discussion related to chunking or changing the type to a 64bit number. https://github.com/rapidsai/cudf/issues/3958

**Comment by GregoryKimball** [comment]
Although [get_arrow.cmake](https://github.com/rapidsai/cudf/blob/branch-22.08/cpp/cmake/thirdparty/get_arrow.cmake) is complex, we won't soon be switching to a new data interface.

**Comment by kkraus14** [comment]
> Although [get_arrow.cmake](https://github.com/rapidsai/cudf/blob/branch-22.08/cpp/cmake/thirdparty/get_arrow.cmake) is complex, we won't soon be switching to a new data interface.

That is a CMake module for finding the Arrow library and linking against it while building libcudf. This issue was discussing the Arrow C Data Interface which is a standard C interface for exchanging Arrow formatted data: https://arrow.apache.org/docs/format/CDataInterface.html.

The two are completely unrelated.

**Comment by vyasr** [comment]
@kkraus14 two years down the line, what is your sense of the viability of switching to use the C Data Interface (full disclosure, I have done no additional research yet :smile:)? It sounds from the tail end of the discussion above that there is a reasonable way for cudf to handle the memory management, and I imagine that the situation has improved since then. If we were to switch, would that provide a sufficient path for us to stop needing to link to arrow? AFAICT we have two main use cases in libcudf:
- [Our interop code reads and writes to arrow tables directly](https://github.com/rapidsai/cudf/blob/branch-23.02/cpp/include/cudf/interop.hpp). This seems like the exact use case for which the data interface exists and could be replaced.
- [We support some form of arrow inputs in our data sources](https://github.com/rapidsai/cudf/blob/branch-23.02/cpp/include/cudf/io/datasource.hpp). I am less sure about whether there is a path to replacing this kind of functionality. @vuule tested a while ago and found that the C++ APIs for creating these datasources was unused, but we still end up with the same issues because our [Python API ends up using the Cython interfaces of Arrow to provide the same sort of functionality](https://github.com/rapidsai/cudf/blob/branch-23.02/python/cudf/cudf/_lib/io/datasource.pyx). Is there a better approach here?

I'm mostly interested in this as a way to simplify managing cudf environments by reducing the strictness of our Arrow version pinnings, which would be nice for users and would make our own maintenance simpler.

CC @trxcllnt in case you have more insight as well.

**Comment by kkraus14** [comment]
> It sounds from the tail end of the discussion above that there is a reasonable way for cudf to handle the memory management, and I imagine that the situation has improved since then.

Yes, it sounds like the memory management aspect of things is very manageable nowadays. We could presumably use pinned host memory as well to speed up the device to host copies (we'd just need to synchronize the stream before returning).

> If we were to switch, would that provide a sufficient path for us to stop needing to link to arrow?

For the cases on interop and being able to import and export Arrow objects, yes. Using C++ there's functions here: https://github.com/apache/arrow/blob/master/cpp/src/arrow/c/bridge.h and for using Cython there's functions here: https://github.com/apache/arrow/blob/6cfd143d344e536978a8721f726415783322bb17/python/pyarrow/includes/libarrow.pxd#L2762-L2789.

Maybe a bit cumbersome for C++ users, but that's the price we have to pay to avoid needing an Arrow linkage unless we wanted to try implementing some `dlopen` / `dlsym` magic.

> * [Our interop code reads and writes to arrow tables directly](https://github.com/rapidsai/cudf/blob/branch-23.02/cpp/include/cudf/interop.hpp). This seems like the exact use case for which the data interface exists and could be replaced.

Yes, with the caveat that we probably should have been using RecordBatches instead of Tables for a while. As linked above, there's helper functions in Arrow for C Data Interface --> RecordBatch and then it should be free to move a RecordBatch into a Table.

> * [We support some form of arrow inputs in our data sources](https://github.com/rapidsai/cudf/blob/branch-23.02/cpp/include/cudf/io/datasource.hpp). I am less sure about whether there is a path to replacing this kind of functionality. @vuule tested a while ago and found that the C++ APIs for creating these datasources was unused, but we still end up with the same issues because our [Python API ends up using the Cython interfaces of Arrow to provide the same sort of functionality](https://github.com/rapidsai/cudf/blob/branch-23.02/python/cudf/cudf/_lib/io/datasource.pyx). Is there a better approach here?

We'd need to remove the `cudf::io::datasource::create` overload that takes an `arrow::io::RandomAccessFile` and the `arrow_io_source` class.

Given the `datasource` base class is designed to be an interface that can be implemented I don't think removing the `create` overload loses much value and is encompassed by `arrow_io_source` anyway. For `arrow_io_source` we'd maybe want to consider rolling that out into its own library or something? Otherwise it should be easy enough to get a buffer from the Arrow RandomAccessFile and feed that buffer into a datasource via something like the `host_buffer` overload?

> I'm mostly interested in this as a way to simplify managing cudf environments by reducing the strictness of our Arrow version pinnings, which would be nice for users and would make our own maintenance simpler. I should say that I don't see much benefit to switching unless doing so removes our libarrow dependence entirely. The benefit (in my view) would come from being able to depend only on pyarrow and not on libarrow.

While I agree this would be nice, I don't know how feasible it is. Given we use the pyarrow Cython pretty extensively, it basically pins us to a major version of pyarrow and therefore libarrow unless we start building a matrix of packages based on arrow version.

It would definitely benefit use cases where only libcudf is being used and not cuDF where they'd no longer need to get libarrow in its entirety.

**Comment by vyasr** [comment]
I also don't know if there's any way to do our Python layer interop with Python-only APIs. I sort of doubt it. Removing the libcudf dependency would at least allow us to get rid of `get_arrow.cmake`, which would be a plus (we would only ever need the branch that executes `pyarrow.get_include`).

**Comment by kkraus14** [comment]
> I also don't know if there's any way to do our Python layer interop with Python-only APIs.

As far as I can tell there is not. Only Cython APIs that are mostly externed from C++ headers.

**Comment by vyasr** [comment]
We are now moving towards implementing APIs using the C data interface. The current work is being tracked in #15193 and is ongoing in #15047, so I'm going to close this issue as the question of interest has been answered.
